modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/TransactionValidator.scala

```diff
@@ -0,0 +1,207 @@
+
+  protected def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def validate(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Signed[Transaction]]] =
+    for {
+      propertiesValidation <- validateTransactionProperties(signedTransaction)
+      balanceValidation <- validateSourceBalances(Seq(signedTransaction.value))
+    } yield propertiesValidation.product(balanceValidation).map(_ => signedTransaction)
+
+  def validate(
+    signedTransactions: Seq[Signed[Transaction]]
+  ): F[ValidationResult[Seq[Signed[Transaction]]]] =
+    for {
+      propertiesValidation <- signedTransactions.traverse(validateTransactionProperties).map(_.sequence)
+      balancesValidation <- validateSourceBalances(signedTransactions.map(_.value))
+    } yield propertiesValidation.product(balancesValidation).map(_ => signedTransactions)
+
+  private def validateTransactionProperties(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Signed[Transaction]]] =
+    for {
+      staticValidation <- Async[F].delay(
+        validateAddressInequality(signedTransaction.value)
+      )
+      signatureValidation <- validateSourceSignature(signedTransaction)
+      sourceValidation <- validateSourceAddress(signedTransaction)
+      lastTxRefValidation <- validateLastTransactionRef(signedTransaction.value)
+    } yield
+      staticValidation
+        .product(signatureValidation)
+        .product(sourceValidation)
+        .product(lastTxRefValidation)
+        .map(_ => signedTransaction)
+
+  private def validateLastTransactionRef(tx: Transaction): F[ValidationResult[Transaction]] =
+    getLastAcceptedTransactionRef(tx.source).map { lastTxRef =>
+      lazy val comparedOrdinals = tx.parent.ordinal.coerce.compare(lastTxRef.ordinal.coerce)
+      lazy val nonZeroOrdinalHasEmptyHash = tx.parent.ordinal.coerce > 0L && tx.parent.hash.coerce.isEmpty
+      lazy val ordinalIsLower = comparedOrdinals < 0L
+      lazy val hashIsNotEqual = comparedOrdinals == 0L && tx.parent != lastTxRef
+
+      if (nonZeroOrdinalHasEmptyHash)
+        NonZeroOrdinalButEmptyHash(tx).invalidNel
+      else if (ordinalIsLower)
+        ParentTxRefOrdinalLowerThenStoredLastTxRef(tx).invalidNel
+      else if (hashIsNotEqual)
+        SameOrdinalButDifferentHashForLastTxRef(tx).invalidNel
+      else
+        tx.validNel
+    }
+
+  private def validateSourceAddress(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Transaction]] =
+    signedTransaction.proofs.traverse {
+      case SignatureProof(id, _) => {
+        for {
+          publicKey <- id.hex.toPublicKey
+          address = publicKey.toAddress
+        } yield address == signedTransaction.value.source
+      }.handleError(_ => false)
+    }.map(_.forall(identity))
+      .ifM(
+        signedTransaction.value.validNel.pure[F],
+        Async[F].pure(SourceAddressAndSignerIdsDontMatch(signedTransaction.value).invalidNel)
+      )
```

Comment:
Eventually:

```scala
signedTransaction.proofs.existsM {
  case SignatureProof(...) => {
     for { ... } yield address != signedTransaction.value.source 
  }.handleError(_ => true)
}.ifM(...)
```

By negating the statement it won't iterate over all the proofs but will stop at the first one which is wrong. Also you get rid of `forall`